### PR TITLE
Optimize varnish ban regexp

### PR DESCRIPTION
**Pull request by bastnic**
| Q             | A
| ------------- | ---
| Branch?       | 2.6
| Bug fix?      | depends on how you treat perf issue
| New feature?  | no
| Deprecations? | no
| License       | MIT

The current ban regexp is not optimize and killed my varnish server to 100% cpu.

Several reasons, but let's explain how that work. This is a question of

> rq/s * ban/s * size of the cache tags header * complexity of the regexp

I've a high number of requests and bans per seconds. The response has a lot of tags.

But I only ban one resource at a time, so the regexp was not that complex at first look.

Varnish stores all ban in a (deduplicated) list. Each cached resource has a pointer to the last ban checked.
If you add bans, when you call a cached resources, it will check against all the new bans and store the last ones.

The number of regexp analysis was high, but I can change my traffic, so I can only adjust the header itself or the regexp.

I looked at the regexp first

> ((^|,)\/pim\/v1\/variants($|,))|((^|,)\/pim\/v1\/variants\/e3615f0c-a957-4a85-88b9-8db30d5c5c02($|,))|((^|,)\/pim\/v1\/models\/bebdee7b-fb53-448b-b79e-1bcf0dbcf050($|,))

Step 1: checking the begining of the resources is not useful if tags are only handled by apip.

> (\/pim\/v1\/variants($|,))|(\/pim\/v1\/variants\/e3615f0c-a957-4a85-88b9-8db30d5c5c02($|,))|(\/pim\/v1\/models\/bebdee7b-fb53-448b-b79e-1bcf0dbcf050($|,))

Stepe 2: regrouping the end tag at the end is better.

> (\/pim\/v1/variants|\/pim\/v1/variants\/e3615f0c-a957-4a85-88b9-8db30d5c5c02|\/pim\/v1/models\/bebdee7b-fb53-448b-b79e-1bcf0dbcf050)($|,)

Step 3: remove the prefix from the regexp

> (/variants|\/variants\/e3615f0c-a957-4a85-88b9-8db30d5c5c02|\/models\/bebdee7b-fb53-448b-b79e-1bcf0dbcf050)($|,)

Step 4: remove the prefix from the source header

Bench:

| subject      | mean    | diff  |
--- | --- | ---
| benchCurrent | 1.612μs | 7.91x |
| benchStep1   | 0.575μs | 2.82x |
| benchStep2   | 0.504μs | 2.47x |
| benchStep3   | 0.492μs | 2.41x |
| benchStep4   | 0.204μs | 1.00x |

This PR only goes to step 2.

On my production, I published Step3 today at 12:15:

image

### CODE REVIEW

File: tests/HttpCache/VarnishPurgerTest.php [hunk 1]
```diff
@@ -118,48 +118,48 @@ public function provideChunkHeaderCases()
...
+            ['(/foo|/bar|/baz)($|\,)'],
         ];
 
         yield 'iris to generate a header with exactly the maximum length and a smaller one' => [
             37,
             ['/foo', '/bar', '/baz'],
             [
-                '((^|\,)/foo($|\,))|((^|\,)/bar($|\,))',
-                '(^|\,)/baz($|\,)',
+                '(/foo|/bar)($|\,)',
```
Comment (julienfalque) [inline tests/HttpCache/VarnishPurgerTest.php:140]: I think something needs to be changed in the splitting logic: the 2 segments would now fit the 37 characters limit if merged.

Comment (bastnic) [reply tests/HttpCache/VarnishPurgerTest.php:140]: hi @julienfalque, I'm fixing tests right now.

**Review by julienfalque — APPROVED** [review]

File: src/HttpCache/VarnishPurger.php [hunk 2]
```diff
@@ -35,7 +36,7 @@ final class VarnishPurger implements PurgerInterface
     public function __construct(array $clients, int $maxHeaderLength = self::DEFAULT_VARNISH_MAX_HEADER_LENGTH)
     {
         $this->clients = $clients;
-        $this->maxHeaderLength = $maxHeaderLength;
+        $this->maxHeaderLength = $maxHeaderLength -\mb_strlen(self::REGEXP_PATTERN) + 2;
```
Comment (dunglas) [inline src/HttpCache/VarnishPurger.php:39]: Missing space?

**Review by dunglas — APPROVED** [review]
Adding support for the xkey module could improve the performance even more.

**Comment by bastnic** [comment]
@dunglas FYI xkey is already abandoned by varnish which prefers ykey https://docs.varnish-software.com/varnish-cache-plus/vmods/ykey/.

Migrating to x/ykey is not "free" as you need the paid version of varnish (maybe I'm wrong on this one).

This PR an optimization for what we already provide in apip.

**Comment by soyuka** [comment]
xkey still works with the free varnish and won't be removed soon

**Review by soyuka — APPROVED** [review]

**Comment by soyuka** [comment]
thanks @bastnic

**Comment by dunglas** [comment]
Yes xkey isn't maintained by Varnish.com (the company) anymore but is still maintained by the community as part of the Varnish modules project (I recently contributed to it and the patch has been merged). 
What I'm suggesting is to support both the current approach and xkey (that have better performance).

**Pull request merged**